### PR TITLE
Database versioning and migration

### LINKED ISSUE #3203: version the database and automatically migrate data
**Issue summary**

Storing data is one of the prime purposes of Forest, and we often play around with different layouts. For example, the copying garbage collector requires a specific layout. The layout will be completely changed by the (not yet finished) non-moving garbage collector. Hubert's work on re-organizing the DB also involves layout changes: #3189

Up until now, we've been okay with deleting the database and starting from a fresh snapshot. This was annoying but acceptable since we never had more than 100GiB of data. Going forward, we'll be setting up nodes with ~14 TiB of data, and we want them to not require any maintenance when new versions of Forest are released.

I propose we put our database in separate folders for each Forest version. If `forest-0.1` has created a database in `%DATA_DIR/0.1/` and we try to load it with `forest-0.2`, these steps should happen:
- `0.1/` is opened and migrated to `pending/`.
- Validity checks are run against `pending/`. We can check that at least 900 tipsets are present, that the network identity (mainnet/calibnet) hasn't changed, and maybe even re-execute a few tipsets.
- If all checks pass (assuming they also passed using the pre-migrated database), atomically rename `pending/` to `0.2` and delete `0.1`.

Note: not all Forest versions require migrations. When no migration is required, the database folder can simply be renamed.

Note: Migrating from version A to version B should apply migrations for all intermediate versions.

Acceptance criteria:
- [X] Per-network versioned database.
- [x] Set up migration system with automated checks for the _pre_ and _post-migration_ databases. If pre-migration checks fail then the post-migration checks can be skipped.
- [ ] Documentation for how to write migrations.
- [X] CI test that goes through all migrations, one by one. This can be done by syncing to calibnet with an old version of Forest. Then sync to calibnet using Forest releases in chronological order.
- [X] CI test that goes through all migrations, from oldest straight to newest. This can be done by first syncing to calibnet with an old version of Forest and then syncing with the newest version.

**Other information and links**

### PR DESCRIPTION
**Pull request by sudo-shashank**
## Summary of changes

Changes introduced in this pull request:

- Implemented database versioning and migration
- Add how to write new db migration doc
- Added CI to test migration by syncing from an older version of forest to latest version of forest
- Use `dev` database if it exist

## Reference issue to close (if applicable)

Closes #3203 

## Change checklist

- [X] I have performed a self-review of my own code,
- [X] I have made corresponding changes to the documentation,
- [X] I have added tests that prove my fix is effective or that my feature works (if possible),
- [X] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/db/migration/mod.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    // Rename db to latest versioned db
+    fs::rename(db_path.as_path(), chain_path(config))?;
+
+    info!("Database Successfully Migrated to {:?}", target_version);
+    Ok(())
+}
+
+// TODO: Add Steps required for new migration
+/// Migrate to an intermediate db version
+fn migrate(_existing_db_path: &Path, next_version: &DBVersion) -> anyhow::Result<()> {
```
Comment (lemmih) [inline src/db/migration/mod.rs:97]: It looks like you want the migrations to happen _inside_ the database folder. That's not safe. If something goes wrong, we want to revert back to the last known good state.

**Review by lemmih — CHANGES_REQUESTED** [review]
You struggle with setting up a framework for database migrations because we have yet to have any database migrations, imo.

Try setting up a database migration from version `Unknown` to version `0.12.0-rc` (unreleased version, you can pretend it exists) to version `dev`. Migrating between three different versions should give you a lot of experience.

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
Database versioning and migration are crucial to having Forest working and resolving issues on running services. This is a feature that needs fancy diagrams and team wide-understanding.

How about you make a proposal (not in the form of code) and present it to the team after daily?

File: src/db/migration/mod.rs (hunk 1 above)
Comment (sudo-shashank) [reply src/db/migration/mod.rs]: I am currently using `temp_dir` for intermediate db migrations, if something goes wrong we can revert back to last good state easily

File: src/db/migration/mod.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    // Run pre-migration checks, which includes:
+    // - re-compute 100 tipsets
+    migration_check(config, &db_path).await?;
+
+    // Iterate over all DBVersion's until database is migrated to lastest version
+    while current_version != target_version {
+        let next_version = match current_version {
+            DBVersion::V0 => DBVersion::V11,
+            _ => break,
+        };
```
[the remark this answers is not in the document]
Comment (sudo-shashank) [reply src/db/migration/mod.rs]: I am maintaining the list of versions now

File: src/db/migration/mod.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        db,
+        Arc::clone(&config.chain),
+        genesis,
+        existing_chain_data_root,
+    )?);
+    let state_manager = Arc::new(StateManager::new(chain_store, Arc::clone(&config.chain))?);
+
+    let ts = state_manager.chain_store().heaviest_tipset();
+    let height = ts.epoch();
+    assert!(height.is_positive());
```
[the remark this answers is not in the document]
Comment (sudo-shashank) [reply src/db/migration/mod.rs]: not required, removed

**Pull request closed without merging**